### PR TITLE
AMMDeposit

### PR DESCRIPTION
**Pull request by florent-uzio**
# Title

This PR aims to:

- Add the AMMDeposit transaction type
- Add a `Flatten` and `Validate` method
- Add all the relevant tests
- Update the `IsAsset` func to handle the case when the Currency is XRP and the issuer is set (invalid)
- Add a test for the new IsAsset check

`AMMDeposit` and `validations_xrpl_objects` have both 100% test coverage.

image

image

### CODE REVIEW

File: xrpl/transaction/amm_deposit.go [hunk 1]
```diff
@@ -1,14 +1,181 @@
...
+
+	if a.Amount2 != nil && a.Amount == nil {
+		return false, errors.New("ammDeposit: must set Amount with Amount2")
+	} else if a.EPrice != nil && a.Amount == nil {
+		return false, errors.New("ammDeposit: must set Amount with EPrice")
+	} else if a.LPTokenOut == nil && a.Amount == nil {
+		return false, errors.New("ammDeposit:  must set at least LPTokenOut or Amount")
+	}
+
+	if ok, err := IsAmount(IsAmountArgs{field: a.Amount, fieldName: "Amount", isFieldRequired: false}); !ok {
```
Comment (GuillemGarciaDev) [inline xrpl/transaction/amm_deposit.go:168]: We should remove the `IsAmountArgs` struct and spread the props. It is much easier to read

**Review by GuillemGarciaDev — CHANGES_REQUESTED** [review]

File: xrpl/transaction/amm_deposit.go (hunk 1 above)
Comment (florent-uzio) [reply xrpl/transaction/amm_deposit.go:168]: I've done the update in #43 , that way I won't have merge conflicts later.

File: xrpl/transaction/validations_xrpl_objects.go [hunk 2]
```diff
@@ -211,6 +211,10 @@ func IsAsset(asset ledger.Asset) (bool, error) {
 		return true, nil
 	}
 
+	if strings.ToUpper(asset.Currency) == "XRP" && asset.Issuer != "" {
```
[the remark this answers is not in the document]
Comment (florent-uzio) [reply xrpl/transaction/validations_xrpl_objects.go:214]: Actually it's not really needed to check if the issuer is a valid address here. If the currency is XRP, the `issuer` field mustn't be undefined, that's why I simply check `!= ""`.

[the remark this answers is not in the document]
Comment (florent-uzio) [reply xrpl/transaction/validations_xrpl_objects.go:214]: Done, I actually made the changes in #43 so I won't have merge conflicts.

File: xrpl/transaction/amm_deposit.go [hunk 3]
```diff
@@ -1,14 +1,181 @@
...
+)
+
+// Deposit funds into an Automated Market Maker (AMM) instance and receive the AMM's liquidity provider tokens (LP Tokens) in exchange.
+// You can deposit one or both of the assets in the AMM's pool.
+// If successful, this transaction creates a trust line to the AMM Account (limit 0) to hold the LP Tokens.
+//
+// Example:
+//
+// ```json
+//
```
[the remark this answers is not in the document]
Comment (florent-uzio) [reply xrpl/transaction/amm_deposit.go:17]: Actually line 17 (empty line) is automatically added by the Golang formatter, it should be fine to leave it like this, no?

File: xrpl/transaction/validations_xrpl_objects.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (GuillemGarciaDev) [reply xrpl/transaction/validations_xrpl_objects.go:214]: True!

**Review by GuillemGarciaDev — APPROVED** [review]

**Pull request merged**